### PR TITLE
Fix bug: markdelete position not moving forward when skipping lost entries

### PR DESCRIPTION
**Pull request by lordcheng10**
### Motivation
discuss thread：https://lists.apache.org/thread/d3f8domf5dnj0h3ktc6sz3kr5tooy4cj

fix not ack for the skiped entry when autoSkipNonRecoverableData=true.

When it was found online that the broker no longer pushed data to the consumer due to loss entry, we tried to turn on the configuration autoSkipNonRecoverableData=true, but found that the message still would not be pushed.
The subscription type here is key share. After adding the log to confirm, it is found that the skipped entry has not been acked, so the mark delete position cannot be moved forward.

**error log:**

00:01:22.623 [BookKeeperClientWorker-OrderedExecutor-33-0] ERROR org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers - [persistent://teg_onion_onion_common_data_gz/teg_onion_onion_common_data_gz/teg_onion_onion_common_data_gz-partition-275 / t_teg_onion_b_teg_onion_onion_common_data_gz_cg_onion_common_data_cos_svr_gz_001] Error reading entries at 15759573:15389 : No such entry, Read Type Normal - Retrying to read in 58.132 seconds

00:01:22.623 [BookKeeperClientWorker-OrderedExecutor-21-0] ERROR org.apache.bookkeeper.client.PendingReadOp - Read of ledger entry failed: L15759573 E15389-E15389, Sent to [****:3181, ***:3181], Heard from [] : bitset = {}, Error = 'No such entry'. First unread entry is (-1, rc = null)

**Through the bookkeeper command, I listed the corresponding entries in the ledger, I found that the smallest entryId is 20153, but the entryId that failed to read is 15389:**
image

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by lordcheng10** [comment]
@hangc0276 @codelipenghui PTAL,thanks!

**Comment by lordcheng10** [comment]
@eolivelli PTAL,thanks!

**Comment by lordcheng10** [comment]
@eolivelli PTAL,thanks!

**Review by eolivelli — CHANGES_REQUESTED** [review]
This is a pretty big change.

Can you please add a test case that reproduces your problem?

As we are adding many new methods to the internal ManagedCursor APi I believe that we should discuss about your approach on the dev mailing list. Can you please start a discussion  there?

**Comment by lordcheng10** [comment]
> This is a pretty big change.
> 
> 
> 
> Can you please add a test case that reproduces your problem?
> 
> 
> 
> As we are adding many new methods to the internal ManagedCursor APi I believe that we should discuss about your approach on the dev mailing list. Can you please start a discussion  there?

OK

**Comment by codelipenghui** [comment]
Yes, this is a big change. It's better to discuss in the dev mail list to make sure everyone is on the same page.

**Comment by lordcheng10** [comment]
>  Yes, this is a big change. It's better to discuss in the dev mail list to make sure everyone is on the same page.

OK  I will  do

**Comment by lordcheng10** [comment]
Here I started a discussion： @eolivelli @codelipenghui 
https://lists.apache.org/list.html?dev@pulsar.apache.org

**Comment by lordcheng10** [comment]
@rdhabalia @merlimat @eolivelli PTAL,thanks!

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java [hunk 1]
```diff
@@ -100,15 +100,16 @@ public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
         } else if (cursor.config.isAutoSkipNonRecoverableData() && exception instanceof NonRecoverableLedgerException) {
             log.warn("[{}][{}] read failed from ledger at position:{} : {}", cursor.ledger.getName(), cursor.getName(),
                     readPosition, exception.getMessage());
-            // try to find and move to next valid ledger
-            final Position nexReadPosition = cursor.getNextLedgerPosition(readPosition.getLedgerId());
+            // try to find and move to next valid entry position
+            PositionImpl nexReadPosition = cursor.getNextAvailablePosition(readPosition);
```
[the remark this answers is not in the document]
Comment (lordcheng10) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java:104]: OK

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Dispatcher.java [hunk 2]
```diff
@@ -44,6 +45,9 @@
 
     boolean canUnsubscribe(Consumer consumer);
 
+    default void registerAutoSkipNonRecoverableDataListener(java.util.function.Consumer<List<Position>> listener) {
```
[the remark this answers is not in the document]
Comment (lordcheng10) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Dispatcher.java:48]: OK

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedCursor.java [hunk 3]
```diff
@@ -74,6 +77,17 @@
      */
     void updateLastActive();
 
+    default void addSkipPosition(Position skipPosition) {
+        // do nothing
+    }
+
+    default List<Position> getSkipPositions() {
+        return null;
```
[the remark this answers is not in the document]
Comment (lordcheng10) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedCursor.java:85]: OK

**Pull request closed without merging**